nntrainer/layers/conv2d_layer.cpp

```diff
@@ -26,30 +26,71 @@
 #include <util_func.h>
 
 namespace nntrainer {
-
-/// @note this will be deleted after conv2d optimization is done!
-#ifdef PROFILE
 namespace {
 
-int pad_profile_key;
-int conv_gemm_profile_key;
-int im2col_key;
-int add_bias_key;
-int clean_up;
-int temp_key;
-
-void register_event() {
-  pad_profile_key = profile::Profiler::Global().registerEvent("zero_pad");
-  im2col_key = profile::Profiler::Global().registerEvent("im2col");
-  conv_gemm_profile_key =
-    profile::Profiler::Global().registerEvent("conv_gemm");
-
-  add_bias_key = profile::Profiler::Global().registerEvent("add_bias_key");
-  clean_up = profile::Profiler::Global().registerEvent("clean_up");
-  temp_key = profile::Profiler::Global().registerEvent("temp_key");
+void col2im(const Tensor &col_matrix, const TensorDim &kdim,
+            const std::array<unsigned, CONV2D_DIM> &padding,
+            const std::array<unsigned, CONV2D_DIM> &mstride,
+            const std::array<unsigned, CONV2D_DIM> &dilation, Tensor &image) {
+  unsigned ph = padding[0];
+  unsigned pw = padding[1];
+
+  unsigned k_height = kdim.height();
+  unsigned k_width = kdim.width();
+
+  /// effective kernel height considering dilation
+  unsigned eff_k_height = (k_height - 1) * dilation[0] + 1;
+  /// effective kernel width considering dilation
+  unsigned eff_k_width = (k_width - 1) * dilation[1] + 1;
+
+  unsigned im_channel = image.channel();
+  int im_height = image.height();
+  int im_width = image.width();
+
+  unsigned hstride = mstride[0];
+  unsigned wstride = mstride[1];
+
+  unsigned hdilation = dilation[0];
+  unsigned wdilation = dilation[1];
+
+  /// image considering padding
+  unsigned im_eff_height = im_height + 2 * ph;
+  unsigned im_eff_width = im_width + 2 * pw;
+  image.setZero();
+
+  int h_stride_end = im_eff_height - eff_k_height - ph;
+  int w_stride_end = im_eff_width - eff_k_width - pw;
```

Comment:
Is it okay just converting an unsigned integer to signed one?
I think those values are supposed to be small enough but this style seems to have potential problems.

Reply:
Thank you for the review!

Those numbers should be at maximum 1000 otherwise the algorithm won't run.
This function is in the critical path, so the indexes are optimized in a way to minimize calculation and would hard to add checks that won't likely to happen.

Although it should be better to be checked beforehand, I added an additional todo in `initialize()` to precalculate if it would overflow.

Reply:
Got it. Thanks for the explanation :)